### DISCUSSION TITLE
Product images are not showing

### DISCUSSION THREAD

**Discussion by venkat402** [bagisto/bagisto#5281] [General]
# Bug report

On production, images are not loading 
I followed the bellow site instructions 
but still not resolved the issue 
https://forums.bagisto.com/topic/188/broken-image-logo-in-bagisto-with-causes-and-solutions

### Issue Description

on landing page 
![image](https://user-images.githubusercontent.com/29934257/73158575-3c938280-410a-11ea-960a-eae7fd518924.png)

same on description page 
![image](https://user-images.githubusercontent.com/29934257/73158380-ac553d80-4109-11ea-834d-56ab14ccc2d2.png)

**Description helps the developers to understand the bug. It describes the problem encountered or some after effect of some kind.**

### Preconditions
**Please provide as detailed information about your environment as possible.**
I run  all the commands bellow but still the issue not fixed at  
composer update

php artisan cache:clear
php artisan view:clear
php artisan route:clear
php artisan clear-compiled
php artisan config:cache

composer dump-autoload
php artisan vendor:publish
php artisan storage:link

rm public/storage 
php artisan storage:link

**Comment by rahulshukla-webkul** [comment]
Hi @venkat402 

Kindly clear your browser cache once, because we can see images - https://prnt.sc/qtdp68 and https://prnt.sc/qtdq4d.

**Comment by venkat402** [comment]
Actually I fixed the issue by removing my custom package from the webkul packages 
When I remove 
"Webkul\\SMS\\": "packages/Webkul/SMS/src"
from psr-4 composer.json then it works fine 
else 
it is failing to show the images

**Comment by nguyenndn** [comment]
![image](https://user-images.githubusercontent.com/17706932/76192224-e428c800-6213-11ea-8c56-8079d622cec4.png)
@rahulshukla-webkul , @jitendra-webkul 
I'm trying a clear cache browser many times, but I still get an error.
Why 'cache/small' or 'cache/original' or 'cache/medium' on URL image product?. I see the image uploaded on the Storage folder. Why we need 'cache/small/product' instead of 'storage/product' ?
Thanks.

**Comment by venkat402** [comment]
Hi,  @keeloren 
 'cache/small' or 'cache/original' or 'cache/medium'
is for minimising image script for different screen lengths.
And getting over to this issue
In root run the bellow command 
php artisan config:cache

**Comment by nguyenndn** [comment]
Hi @venkat402 , I'm using docker. 
run php artisan config:cache.
![image](https://user-images.githubusercontent.com/17706932/76200208-ce230380-6223-11ea-8cfe-63c2b60b8b21.png)

but still gets an error.

![image](https://user-images.githubusercontent.com/17706932/76200264-e3982d80-6223-11ea-900e-f74417748b14.png)

where is  'cache/small' and why it includes to URL image :(

**Comment by rahulshukla-webkul** [comment]
Hi @keeloren 

Please follow the given link for solution - https://forums.bagisto.com/topic/188/broken-image-logo-in-bagisto-with-causes-and-solutions

**Comment by nguyenndn** [comment]
HI @rahulshukla-webkul , I try to run that command on the link solution. But it does not work :((

**Comment by rahulshukla-webkul** [comment]
Hi @keeloren 

Do the following things, from the root of your directory - 

`cd public`

`unlink storage`

`cd ..`

`php artisan storage:link`

Thanks

**Comment by nguyenndn** [comment]
![image](https://user-images.githubusercontent.com/17706932/76201383-c19faa80-6225-11ea-8cfa-7d044d1153cd.png)
Hi @rahulshukla-webkul,
I follow this, but nothing has changed. So sad ~.~
I spent a lot of time on this and wish to fix this. Thank you.

**Comment by nguyenndn** [comment]
![image](https://user-images.githubusercontent.com/17706932/76201872-8a7dc900-6226-11ea-9acc-13a635542bea.png)
In the Storage folder, I do not see the folder 'cache', or 'large'. But why 'cache/large' have been include URL image?
![image](https://user-images.githubusercontent.com/17706932/76202019-c31da280-6226-11ea-804b-c9d844212250.png)

**Comment by rahulshukla-webkul** [comment]
Hi @venkat402 

Actually we are caching images in three different sizes.

Can you tell me , whats is your APP_URL in .env file ?

Is that - localhost::9000 or not.

Thanks

**Comment by nguyenndn** [comment]
HI @rahulshukla-webkul @venkat402 .

My APP_URL = http://localhost:9000

Please tell me to know where the folder has caching images in three different sizes? Thanks.

**Comment by venkat402** [comment]
Change APP_URL =http://localhost/project_name/public/

**Comment by nguyenndn** [comment]
Hi @venkat402 @rahulshukla-webkul 
I solved it. Many thank for support :)

**Comment by MothanaDo** [comment]
> Hi @venkat402 @rahulshukla-webkul
> I solved it. Many thank for support :)

how u r solve it :(

**Comment by MothanaDo** [comment]
**when i run : composer dump-autoload**

Generating optimized autoload files
Deprecation Notice: Class Webkul\Admin\DataGrids\CartRuleCouponsDataGrid located in ./packages/Webkul/Admin/src/DataGrids/CouponsDataGrid.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Shop\TriggerCest located in ./tests/trigger/Shop/TriggerCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Shop\DatabaseLogicCest located in ./tests/trigger/Shop/DatabaseLogicCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Webkul\Admin\ProductCest located in ./tests/functional/Admin/Catalog/ProductCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Catalog\AttributeFamilyCest located in ./tests/functional/Admin/Catalog/AttributeFamilyCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Catalog\AttributeCest located in ./tests/functional/Admin/Catalog/AttributeCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Catalog\CategoryCest located in ./tests/functional/Admin/Catalog/CategoryCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Sales\InvoiceCest located in ./tests/functional/Admin/Sales/InvoiceCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Sales\RefundCest located in ./tests/functional/Admin/Sales/RefundCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Sales\OrderCest located in ./tests/functional/Admin/Sales/OrderCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Sales\ShipmentsCest located in ./tests/functional/Admin/Sales/ShipmentsCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Customer\GroupsCest located in ./tests/functional/Admin/Customer/GroupsCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Customer\CustomerCest located in ./tests/functional/Admin/Customer/CustomerCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Customer\NewsletterSubscriptionCest located in ./tests/functional/Admin/Customer/NewsletterSubscriptionCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Admin\Customer\ReviewCest located in ./tests/functional/Admin/Customer/ReviewCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Shop\ProductControllerCest located in ./tests/functional/Shop/ProductControllerCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Shop\GuestCheckoutCest located in ./tests/functional/Shop/GuestCheckoutCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Cart\CartTaxesCest located in ./tests/functional/Shop/CartTaxesCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Checkout\Cart\CartCest located in ./tests/functional/Checkout/Cart/CartCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Functional\Checkout\Cart\OrderCest located in ./tests/functional/Checkout/Order/OrderCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Acceptance\GuestCheckoutCest located in ./tests/acceptance/GuestCheckoutCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Category\CategoryCest located in ./tests/unit/Category/CategoryCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Core\CoreCest located in ./tests/unit/Core/CoreCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Product\Helpers\ProductTypeCest located in ./tests/unit/Product/Helpers/ProductTypeCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Checkout\Cart\Models\CartModelCest located in ./tests/unit/Checkout/Cart/Models/CartModelCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Checkout\Cart\Controllers\CartControllerCest located in ./tests/unit/Checkout/Cart/Controllers/CartControllerCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
Deprecation Notice: Class Tests\Unit\Tax\Helpers\TaxCest located in ./tests/unit/Tax/Helpers/TaxCest.php does not comply with psr-4 autoloading standard. It will not autoload anymore in Composer v2.0. in phar:///usr/local/bin/composer/src/Composer/Autoload/ClassMapGenerator.php:201
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover

In PackageManifest.php line 177:

  The /home/bagisto/public_html/bagisto/bootstrap/cache directory must be pre
  sent and writable.


Script @php artisan package:discover handling the post-autoload-dump event returned with error code 1

**Comment by riwash** [comment]
make ur env url right like this APP_URL=http://localhost/project/multi-use/sham-ecommerce/public/

**Comment by sebastien15** [comment]
the other thing that helped me through my three days journey is to manually delete storage folder in public folder and then do php artisan storage:link

**Comment by vivekshubhu** [comment]
@venkat402 Actually I've got the same problem. I run all the commands above and also changed app_url as above. but it does not work.

![1](https://user-images.githubusercontent.com/51329802/82808553-ec56e200-9ea9-11ea-9332-d01635552f15.PNG)

**Reply by MusheAbdulHakim** [reply]
Were you able to fix this?. I have the same issue. I have try clearing all the cache and deleting and linking the storage folder. Nothing has change. The images seem to be displaying fine at the admin panel though.

**Reply by MusheAbdulHakim** [reply]
I have also set my APP_URL to point to the public folder.

**Comment by venkat402** [comment]
Delete storage directory resides in the public folder
Then create storage link

On Mon, May 25, 2020 at 4:52 PM vivekshubhu <notifications@github.com>
wrote:

> @venkat402 <https://github.com/venkat402> Actually I've got the same
> problem. I run all the commands above and also changed app_url as above.
> but it does not work.
>
> [image: 1]
> <https://user-images.githubusercontent.com/51329802/82808553-ec56e200-9ea9-11ea-9332-d01635552f15.PNG>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bagisto/bagisto/issues/2152#issuecomment-633523544>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHEMFMIGIBPEHEZHU7WCLZLRTJIGLANCNFSM4KL5JMKQ>
> .
>


-- 
Thanks and Regards
Venkat A
P.no : 91+ 9052 380 780.
Email: adapala.vr@gmail.com

**Comment by farshadff** [comment]
any fix for this ?? when moving image files from one host to another this happens i did all above but still the images wont load they load in admin panel but not front shop

**Comment by riwash** [comment]
php artisan storage:link this every time if u move folder it work or u can change upload image file location to public

**Comment by farshadff** [comment]
@riwash  no i tried that this didnt solve it for me

**Comment by chiendv** [comment]
Hi,
I get stuck in this error when running Bagisto with nginx and reserve-proxy (client -> proxy(port 80) -> nginx (port 83) )

Nginx is running at port 83 and root is point to bagisto public directory.

APP_URL=http://domain.com

php artisan storage:link said
The [/var/www/html/xxx/public/storage] link has been connected to [/var/www/html/xxx/storage/app/public].

images are good in admin but not show in frontend.

any help please!

**Comment by devansh-webkul** [comment]
Hi @chiendv,

There are possibly only two reasons,

- First, either your `APP_URL` is not correct in `.env` file. Just change that and run `php artisan optimize`.
- Second, the GD extension maybe not be installed as mentioned in the docs because in Bagisto we are using the Laravel intervention package which optimizes the image.
https://devdocs.bagisto.com/1.x/introduction/requirements.html#php-extensions

**Comment by chiendv** [comment]
Hi @devansh-webkul,
Thanks for you quick response. 

I has resolved it by check config of nginx and comment out config lines for cache static files
```
        #location ~* \.(jpg|jpeg|png|gif|ico|mp3|mp4|mov)$ {
        #       expires 30d;
        #}
```
Hope it can help other people with this kind of issue.

**Comment by merdiano** [comment]
> Hi @devansh-webkul,
> Thanks for you quick response.
> 
> I has resolved it by check config of nginx and comment out config lines for cache static files
> 
> ```
>         #location ~* \.(jpg|jpeg|png|gif|ico|mp3|mp4|mov)$ {
>         #       expires 30d;
>         #}
> ```
> 
> Hope it can help other people with this kind of issue.

what if i want to cache statick files?

**Comment by devansh-webkul** [comment]
We are using the default Laravel configuration and mostly these are cached. If you need to customize you can feel free to do it at your instance.

**Comment by williamagyapong** [comment]
I was happy when I found this discussion after running into the same issue of products images not showing, however, none of the suggestions given here does seem to work for me. Any other help will be greatly appreciated. Below is a screenshot:

<img width="1675" alt="Screen Shot 2021-10-27 at 3 15 20 PM" src="https://user-images.githubusercontent.com/15948490/139148971-9de5e4a3-ad2d-4bfb-b95c-85c1f76f62e8.png">

**Reply by chiendv** [reply]
@williamagyapong 
First of all, check and make sure that upload process is working well and your files are uploaded to storage directory. 

Second step is check for configs of nginx and try to comment out unnecessary lines and reload nginx

Last step is follow document to re-link storage directory.

**Reply by williamagyapong** [reply]
Thank you for your suggestions. Everything started working fine after I followed the installation guide [here](https://devdocs.bagisto.com/1.x/introduction/installation.html) to install a new instance of the bagisto app.

**Comment by amomad99** [comment]
I got this error, and there was no solution at all.
I tried to solve this the whole day.

To solve it, go to your .ENV file, and in APP_URL, add the URL and port like this (APP_URL=http://localhost:8000)
it worked for me...

Thanks

**Reply by Sanjaybhattwebkul** [reply]
Hi there,
Pease raise a ticket [Here](https://bagisto.uvdesk.com/en/customer/create-ticket/), i will guide you over the ticket. i need to check on your instance.

**Reply by amomad99** [reply]
Nah! I solved that problem. Thank you...

**Reply by Sanjaybhattwebkul** [reply]
Hi there,
**Thanks** for update, if still you have any **query** then please **let me know** and feel free to raise a support **ticket**.

**Comment by MusheAbdulHakim** [comment]
Is there any real answer to this issue at all?. I have try all the above and the images are still not showing in the frontend side. but is displaying fine in the admin panel.
``APP_URL=http://localhost/mushe-commerce/public``
I have run the following commands: 
```
php artisan config:clear
php artisan view:clear
php artisan route:clear
php artisan cache:clear
```
I have also try deleting the storage folder in the public folder and relink the folder with 
`` php artisan storage:link``
Still nothing has change.

![Screenshot at 2022-12-15 19-19-18](https://user-images.githubusercontent.com/58814410/207948844-eb568d74-2149-4941-a893-2d618524eb85.png)


Any ideas on how to fix this.

**Comment by Sanjaybhattwebkul** [comment]
Hi there, Please follow the below steps for the resolved the issue.

**1.** cd public
**2.** unlink storage
**3.** cd ..
**4.** php artisan storage:link
**5.**  Remove the cache folder inside the bootstrap in your root directory.
**6.**  Create a cache folder inside the  bootstrap in your root directory.
**7.** php srtisan vendor:publish --force 
**8.** Press [27]  and Enter.
**9.** Go to velocity and refresh the page.

**Reply by MusheAbdulHakim** [reply]
I am still having the same issue.

**Comment by Sanjaybhattwebkul** [comment]
Hi there,
Make sure the **APP_URL** should be correct in your **.env** file.

**Reply by MusheAbdulHakim** [reply]
I have switch between different options for the APP_URL value but it didn't work either

**Comment by Sanjaybhattwebkul** [comment]
Hi there,
Please raise a ticket  [Here](https://bagisto.uvdesk.com/en/customer/create-ticket/), i need to check on your instance.

**Comment by Davronjuraev** [comment]
Hi, guys
i see, this issue not closed yet.
if all above solution not worked try this one:

GD Library extension not available with this PHP installation - thats why is not showing. 
You need to enable extansion.

using xampp apache server.

Go to php folder in xampp and open the php.ini and php configurations settings file, and change the line ;extension=gd2 or the line ;extension=gd to just extension=gd2

For Ubuntu you should install it manually:

PHP7.4: sudo apt-get install php7.4-gd
PHP8.0: sudo apt-get install php8.0-gd
PHP8.1: sudo apt-get install php8.1-gd

Others are same to installation or Enable.

Restart all your Services like Apache etc.

Good Luck. And thanks for Developers Bagisto

**Comment by devansh-webkul** [comment]
Please refer here,
https://devdocs.bagisto.com/1.5.x/introduction/requirements.html#php-extensions

**Comment by sudeep2003** [comment]
The image is not showing when I am working on it locally with a production database (remotely accessed database).
